File: Claude/Simples/bfs/Bfs.java
// Input: Using the previous graph, how to implement the Breadth-First Search algorithm?

import java.util.*;

public class Bfs<T> {
    private Map<T, List<T>> adjacencyList;

    public Bfs() {
        adjacencyList = new HashMap<>();
    }

    public void addEdge(T source, T destination) {
        if (!adjacencyList.containsKey(source)) {
            addVertex(source);
        }
        if (!adjacencyList.containsKey(destination)) {
            addVertex(destination);
        }
        adjacencyList.get(source).add(destination);
    }

    public void addVertex(T vertex) {
        adjacencyList.putIfAbsent(vertex, new ArrayList<>());
    }
    
    public List<T> getNeighbors(T vertex) {
        return adjacencyList.get(vertex);
    }

    public List<T> breadthFirstSearch(T startVertex) {
        Set<T> visited = new HashSet<>();
        List<T> result = new ArrayList<>();
        Queue<T> queue = new LinkedList<>();

        visited.add(startVertex);
        queue.offer(startVertex);
        result.add(startVertex);

        while (!queue.isEmpty()) {
            T currentVertex = queue.poll();
            for (T neighbor : getNeighbors(currentVertex)) {
                if (!visited.contains(neighbor)) {
                    visited.add(neighbor);
                    queue.offer(neighbor);
                    result.add(neighbor);
                }
            }
        }

        return result;
    }
        public static void main(String[] args) {
        Bfs<String> g = new Bfs<>();

        g.addEdge("A", "B"); 
        g.addEdge("A", "C"); 
        g.addEdge("C", "D");

        List<String> bfs = g.breadthFirstSearch("A");

        System.out.println("" + bfs);
    }
}

/*
Score: 1
*/

File: Claude/Simples/bfs/bfs.py
#Input: Using the previous graph, how to implement the Breadth-First Search algorithm?

from collections import deque

class Graph:
    def __init__(self):
        self.adjacency_list = {}

    def add_vertex(self, vertex):
        if vertex not in self.adjacency_list:
            self.adjacency_list[vertex] = []

    def add_edge(self, vertex1, vertex2):
        if vertex1 in self.adjacency_list and vertex2 in self.adjacency_list:
            self.adjacency_list[vertex1].append(vertex2)
            self.adjacency_list[vertex2].append(vertex1)

    def print_graph(self):
        for vertex in self.adjacency_list:
            print(vertex, ":", self.adjacency_list[vertex])

    def bfs(self, start_vertex):
        visited = set()
        queue = deque([start_vertex])

        while queue:
            vertex = queue.popleft()
            if vertex not in visited:
                visited.add(vertex)
                print(vertex)
                for neighbor in self.adjacency_list[vertex]:
                    queue.append(neighbor)

# Example usage
graph = Graph()
graph.add_vertex('A')
graph.add_vertex('B')
graph.add_vertex('C')
graph.add_vertex('D')
graph.add_vertex('E')
graph.add_edge('A', 'B')
graph.add_edge('A', 'C')
graph.add_edge('B', 'D')
graph.add_edge('D', 'E')

print("Breadth-First Search (BFS) traversal:")
graph.bfs('A')

"""
Score: 1
"""